x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/search/PITAwareQueryClient.java

```diff
@@ -0,0 +1,89 @@
+import org.elasticsearch.xpack.core.search.action.ClosePointInTimeAction;
+import org.elasticsearch.xpack.core.search.action.ClosePointInTimeRequest;
+import org.elasticsearch.xpack.core.search.action.ClosePointInTimeResponse;
+import org.elasticsearch.xpack.core.search.action.OpenPointInTimeAction;
+import org.elasticsearch.xpack.core.search.action.OpenPointInTimeRequest;
+import org.elasticsearch.xpack.core.search.action.OpenPointInTimeResponse;
+import org.elasticsearch.xpack.eql.session.EqlSession;
+import org.elasticsearch.xpack.ql.index.IndexResolver;
+
+import static org.elasticsearch.xpack.ql.util.ActionListeners.map;
+import static org.elasticsearch.action.ActionListener.wrap;
+
+/**
+ * Extension of basic query, adding awareness Point-in-Time awareness.
+ * Opens a point-in-time, uses it for all queries and closes it when disposed,
+ * freeing consumer from doing any special management for it.
+ */
+public class PITAwareQueryClient extends BasicQueryClient {
+
+    private String pitId;
+    private final TimeValue keepAlive;
+
+    public PITAwareQueryClient(EqlSession eqlSession) {
+        super(eqlSession);
+        this.keepAlive = eqlSession.configuration().requestTimeout();
+    }
+
+    @Override
+    protected void search(SearchRequest search, ActionListener<SearchResponse> listener) {
+        // no pid, ask for one
+        if (pitId == null) {
+            openPIT(wrap(r -> {
+                pitId = r;
+                searchWithPIT(search, listener);
+            }, listener::onFailure));
+        }
+        else {
+            searchWithPIT(search, listener);
+        }
+    }
+
+    private void searchWithPIT(SearchRequest search, ActionListener<SearchResponse> listener) {
+        // don't increase the keep alive
+        search.source().pointInTimeBuilder(new SearchSourceBuilder.PointInTimeBuilder(pitId, null));
+        // get the pid on each request
+        super.search(search, wrap(r -> {
+                pitId = r.pointInTimeId();
+                listener.onResponse(r);
+            },
+            // always close PIT in case of exceptions
+            e -> {
+                if (pitId != null) {
+                    close(wrap(b -> {}, listener::onFailure));
+                }
+                listener.onFailure(e);
+            }));
+    }
+
+    private void openPIT(ActionListener<String> listener) {
+        OpenPointInTimeRequest request = new OpenPointInTimeRequest(
+            indices,
+            IndexResolver.FIELD_CAPS_INDICES_OPTIONS,
+            keepAlive,
+            null,
+            null
+        );
+        client.execute(OpenPointInTimeAction.INSTANCE, request, map(listener, OpenPointInTimeResponse::getSearchContextId));
+    }
+
+    @Override
+    public void close(ActionListener<Boolean> listener)  {
+        client.execute(ClosePointInTimeAction.INSTANCE, new ClosePointInTimeRequest(pitId),
+            map(listener, ClosePointInTimeResponse::isSucceeded));
+        pitId = null;
```

Comment:
It would be nice to have tests that check that all PITs are closed when the request finishes ?

Reply:
Indeed. I've checked and found the assertion for no search contexts existed only in SQL so I moved it into QL to be reused into EQL as well. Commented out close to see the assertion being tripped.
I believe that's enough.

Reply:
agreed, thanks